File: transpeter/parser.py
from utils import TokenType, Peekable, CompilerError
import astnode as ast


class ParserError(CompilerError):
    pass


class Parser:
    __slots__ = 'tokens'

    EOF = 'eof'

    def __init__(self, tokens):
        self.tokens = Peekable(tokens, 'eof')

    def expect(self, token):
        """
        Consumes the next token and raises a `ParserError` if it is not equal to `token`.
        """
        next_token = self.tokens.next()
        if next_token == Parser.EOF:
            raise ParserError(f'unexpected EOF, expected \'{token}\'')
        elif next_token != token:
            raise ParserError(f'line {next_token.line}: expected \'{token}\', got \'{next_token.value}\'')
        return next_token

    def parse(self, program) -> ast.Program:
        instructions = []
        while self.tokens.peek != Parser.EOF:
            if self.tokens.peek == TokenType.TYPE:
                # function or declaration
                var_type, name, line = self.parse_declaration()
                if self.tokens.peek == '=':
                    # declaration with initialization
                    self.tokens.next()
                    expr = self.parse_expression()
                    self.expect(';')
                    instructions.append(ast.Declaration(line, var_type, name, expr))
                elif self.tokens.peek == ';':
                    # declaration without initialization
                    self.tokens.next()
                    instructions.append(ast.Declaration(line, var_type, name))
                else:
                    # function
                    args, block = self.parse_func()
                    instructions.append(ast.Func(line, var_type, name, args, block))
            else:
                # statement
                instructions.append(self.parse_statement())
        return ast.Program(program, instructions)

    # returns args, statement (temp?)
    def parse_func(self):
        self.expect('(')
        args = []
        if self.tokens.peek != ')':
            arg_type = self.expect(TokenType.TYPE)
            name = self.expect(TokenType.IDENTIFIER)
            args.append(ast.Declaration(arg_type.line, arg_type.value, name.value))
        while self.tokens.peek != ')':
            self.expect(',')
            arg_type = self.expect(TokenType.TYPE)
            name = self.expect(TokenType.IDENTIFIER)
            args.append(ast.Declaration(arg_type.line, arg_type.value, name.value))
        self.tokens.next()
        block = self.parse_block()
        return args, block

    def parse_block(self) -> ast.Block:
        """
        Parses a block.
        """
        line = self.expect('{').line
        block = []
        while self.tokens.peek != '}':
            if self.tokens.peek == TokenType.TYPE:
                # declaration
                var_type, name, line = self.parse_declaration()
                if self.tokens.peek == '=':
                    self.tokens.next()
                    expr = self.parse_expression()
                    self.expect(';')
                    block.append(ast.Declaration(line, var_type, name, expr))
                else:
                    self.expect(';')
                    block.append(ast.Declaration(line, var_type, name))
            else:
                # statement
                block.append(self.parse_statement())
        self.tokens.next()
        return ast.Block(line, block)

    def parse_declaration(self):
        """
        Parses a declaration.
        """
        next_token = self.expect(TokenType.TYPE)
        name = self.expect(TokenType.IDENTIFIER).value
        return next_token.value, name, next_token.line

    def parse_statement(self):
        """
        Parses a statement.
        """
        if self.tokens.peek == '{':  # block
            return self.parse_block()
        elif self.tokens.peek in (TokenType.IF, TokenType.REPEAT, TokenType.WHILE):
            # repeat, while, if-else
            ctrl = self.tokens.next()
            line = ctrl.line
            self.expect('(')
            expr = self.parse_expression()
            self.expect(')')
            statement = self.parse_statement()
            if ctrl == TokenType.IF:
                if self.tokens.peek == TokenType.ELSE:
                    self.tokens.next()
                    else_statement = self.parse_statement()
                    return ast.If(line, expr, statement, else_statement)
                else:
                    return ast.If(line, expr, statement)
            elif ctrl == TokenType.WHILE:
                return ast.While(line, expr, statement)
            else:
                assert ctrl == TokenType.REPEAT
                return ast.Repeat(line, expr, statement)
        elif self.tokens.peek == TokenType.RETURN:
            # return
            line = self.tokens.next().line
            expr = self.parse_expression()
            self.expect(';')
            return ast.Return(line, expr)
        elif self.tokens.peek == TokenType.INLINE:
            # inline
            inline = self.tokens.next()
            return ast.Inline(inline.line, inline.value)
        elif self.tokens.peek == TokenType.IDENTIFIER:
            name = self.tokens.next()
            if self.tokens.peek == '(':
                # function call
                args = self.parse_func_call()
                self.expect(';')
                return ast.FuncCall(name.line, name.value, args)
            next_token = self.tokens.next()
            if next_token in ('=', '+=', '-=', '*=', '/=', '%='):
                # assignment
                assign_op = next_token.value
                expr = self.parse_expression()
                self.expect(';')
                return ast.Assign(name.line, assign_op, name.value, expr)
            elif next_token == Parser.EOF:
                raise ParserError('unexpected EOF')
            else:
                raise ParserError('line {}: expected function call or assignment'.format(next_token.line))
        elif self.tokens.peek == Parser.EOF:
            raise ParserError('unexpected EOF')
        else:
            raise ParserError('line {}: unexpected token: \'{}\''.format(self.tokens.peek.line, self.tokens.peek.value))

    def parse_expression(self):
        """
        Parses an expression.
        """
        expr = self.parse_and()
        while self.tokens.peek == 'or':
            line = self.tokens.next().line
            next_expr = self.parse_and()
            expr = ast.BinOp(line, 'or', expr, next_expr)
        return expr

    def parse_and(self):
        expr = self.parse_not()
        while self.tokens.peek == 'and':
            line = self.tokens.next().line
            next_expr = self.parse_not()
            expr = ast.BinOp(line, 'and', expr, next_expr)
        return expr

    def parse_not(self):
        if self.tokens.peek == "not":
            line = self.tokens.next().line
            return ast.UnOp(line, "not", self.parse_equality())
        return self.parse_equality()

    def parse_equality(self):
        expr = self.parse_relational()
        while self.tokens.peek in ('==', '!='):
            op = self.tokens.next()
            next_expr = self.parse_relational()
            expr = ast.BinOp(op.line, op.value, expr, next_expr)
        return expr

    def parse_relational(self):
        expr = self.parse_additive()
        while self.tokens.peek in ('<', '>', '<=', '>='):
            op = self.tokens.next()
            next_expr = self.parse_additive()
            expr = ast.BinOp(op.line, op.value, expr, next_expr)
        return expr

    def parse_additive(self):
        expr = self.parse_term()
        while self.tokens.peek in ('+', '-'):
            op = self.tokens.next()
            next_expr = self.parse_term()
            expr = ast.BinOp(op.line, op.value, expr, next_expr)
        return expr

    def parse_term(self):
        expr = self.parse_factor()
        while self.tokens.peek in ('*', '/', '%'):
            op = self.tokens.next()
            next_expr = self.parse_factor()
            expr = ast.BinOp(op.line, op.value, expr, next_expr)
        return expr

    def parse_factor(self):
        if self.tokens.peek == TokenType.IDENTIFIER:
            # function call or variable access
            next_token = self.tokens.next()
            if self.tokens.peek == '(':
                args = self.parse_func_call()
                return ast.FuncCall(next_token.line, next_token.value, args)
            return ast.Var(next_token.line, next_token.value)
        elif self.tokens.peek == '(':
            # parenthesis
            self.tokens.next()
            expr = self.parse_expression()
            self.expect(')')
            return expr
        elif self.tokens.peek in ('-', '+'):
            # unary operator
            op = self.tokens.next()
            expr = self.parse_factor()
            return ast.UnOp(op.line, op.value, expr)
        elif self.tokens.peek == TokenType.INT:
            # integer literal
            val = self.tokens.next()
            return ast.Int(val.line, val.value)
        elif self.tokens.peek == Parser.EOF:
            raise ParserError('unexpected EOF')
        else:
            raise ParserError('line {}: unexpected token: \'{}\''.format(self.tokens.peek.line, self.tokens.peek.value))

    # returns no node, only the arguments
    def parse_func_call(self):
        self.expect('(')
        args = []
        if self.tokens.peek != ')':
            args.append(self.parse_expression())
        while self.tokens.peek != ')':
            self.expect(',')
            args.append(self.parse_expression())
        self.tokens.next()
        return args

File: transpeter-rs/src/parser.rs
//! The parser.

use std::fmt;
use std::iter::Peekable;

use crate::ast::*;
use crate::token::{Token, TokenKind};
use crate::util::{CompilerError, CompilerResult};

fn binary_bp(op: &TokenKind) -> Option<(BinaryOpKind, (u8, u8))> {
    use BinaryOpKind::*;

    Some(match op {
        TokenKind::Plus => (Plus, (9, 10)),
        TokenKind::Minus => (Minus, (9, 10)),
        TokenKind::Star => (Star, (11, 12)),
        TokenKind::Slash => (Slash, (11, 12)),
        TokenKind::Percent => (Percent, (11, 12)),
        TokenKind::EqEq => (EqEq, (7, 8)),
        TokenKind::NotEq => (NotEq, (7, 8)),
        TokenKind::Greater => (Greater, (7, 8)),
        TokenKind::GreaterEq => (GreaterEq, (7, 8)),
        TokenKind::Less => (Less, (7, 8)),
        TokenKind::LessEq => (LessEq, (7, 8)),
        TokenKind::And => (And, (3, 4)),
        TokenKind::Or => (Or, (1, 2)),
        _ => return None,
    })
}

fn unary_bp(op: &TokenKind) -> Option<(UnaryOpKind, u8)> {
    use UnaryOpKind::*;

    Some(match op {
        TokenKind::Plus => (Plus, 13),
        TokenKind::Minus => (Minus, 13),
        TokenKind::Not => (Not, 5),
        _ => return None,
    })
}

fn expected(msg: impl fmt::Display, token: Token) -> CompilerError {
    CompilerError::new(format!("expected {}, got {}", msg, token.kind), token.pos)
}

/// The parser state.
pub struct Parser<I: Iterator<Item = Token>> {
    iter: Peekable<I>,
}

impl<I: Iterator<Item = Token>> Parser<I> {
    /// Creates a new `Parser` from a `Token` iterator.
    pub fn new(tokens: I) -> Self {
        Self {
            iter: tokens.peekable(),
        }
    }

    /// Consumes and returns the next `Token`.
    fn next(&mut self) -> Token {
        self.iter.next().expect("called `Parser::next` after EOF")
    }

    /// Returns the next `Token` without consuming it.
    fn peek(&mut self) -> &Token {
        self.iter.peek().expect("called `Parser::peek` after EOF")
    }

    fn expect(&mut self, tk: &TokenKind) -> CompilerResult<Token> {
        let token = self.next();
        if &token.kind == tk {
            Ok(token)
        } else {
            Err(expected(tk, token))
        }
    }

    fn expect_identifier(&mut self) -> CompilerResult<Ident> {
        let token = self.next();
        let pos = token.pos;
        match token.kind {
            TokenKind::Identifier(name) => Ok(Ident { pos, name }),
            _ => Err(expected("identifier", token)),
        }
    }

    fn expect_type(&mut self) -> CompilerResult<Type> {
        let token = self.next();
        let pos = token.pos;
        match token.kind {
            TokenKind::Type(name) => Ok(Type { pos, name }),
            _ => Err(expected("type", token)),
        }
    }

    fn optional(&mut self, tk: &TokenKind) -> bool {
        if &self.peek().kind == tk {
            self.next();
            true
        } else {
            false
        }
    }

    pub fn parse_program(&mut self, name: &str) -> CompilerResult<Program> {
        let mut items = Vec::new();
        loop {
            let item = match self.peek().kind {
                TokenKind::Eof => break,
                TokenKind::Type(_) => {
                    let mut decl = self.parse_declaration()?;
                    let token = self.next();
                    match token.kind {
                        TokenKind::LeftParen => {
                            // function definition
                            let parameters =
                                self.parse_list(Self::parse_declaration, &TokenKind::RightParen)?;
                            let statements = self.parse_block()?;
                            Item::Function {
                                name: decl.name,
                                return_type: decl.type_,
                                parameters,
                                statements,
                            }
                        }
                        TokenKind::Eq => {
                            // declaration with initialization
                            decl.init = Some(self.parse_expr()?);
                            self.expect(&TokenKind::Semicolon)?;
                            Item::Statement(Statement::Declaration(decl))
                        }
                        TokenKind::Semicolon => Item::Statement(Statement::Declaration(decl)),
                        _ => return Err(expected("function definition or declaration", token)),
                    }
                }
                _ => Item::Statement(self.parse_statement()?),
            };
            items.push(item);
        }
        let name = name.to_string();
        Ok(Program { items, name })
    }

    fn parse_list<T>(
        &mut self,
        p: impl Fn(&mut Self) -> CompilerResult<T>,
        end: &TokenKind,
    ) -> CompilerResult<Vec<T>> {
        let mut elems = Vec::new();
        while &self.peek().kind != end {
            elems.push(p(self)?);
            if !self.optional(&TokenKind::Comma) {
                break;
            }
        }
        self.expect(end)?;
        Ok(elems)
    }

    fn parse_declaration(&mut self) -> CompilerResult<Declaration> {
        let type_ = self.expect_type()?;
        let name = self.expect_identifier()?;
        let init = None;
        Ok(Declaration { type_, name, init })
    }

    fn parse_block(&mut self) -> CompilerResult<Vec<Statement>> {
        self.expect(&TokenKind::LeftBrace)?;
        let mut statements = Vec::new();
        while self.peek().kind != TokenKind::RightBrace {
            statements.push(self.parse_statement()?);
        }
        self.next(); // has kind TokenKind::RightBrace
        Ok(statements)
    }

    fn parse_statement(&mut self) -> CompilerResult<Statement> {
        let token = self.peek();
        let pos = token.pos;
        Ok(match token.kind {
            TokenKind::LeftBrace => Statement::Block {
                statements: self.parse_block()?,
            },
            TokenKind::If => {
                self.next();
                self.expect(&TokenKind::LeftParen)?;
                let condition = self.parse_expr()?;
                self.expect(&TokenKind::RightParen)?;
                let if_statement = Box::new(self.parse_statement()?);
                let else_statement = if self.optional(&TokenKind::Else) {
                    Some(Box::new(self.parse_statement()?))
                } else {
                    None
                };
                Statement::If {
                    pos,
                    condition,
                    if_statement,
                    else_statement,
                }
            }
            TokenKind::While => {
                self.next();
                self.expect(&TokenKind::LeftParen)?;
                let condition = self.parse_expr()?;
                self.expect(&TokenKind::RightParen)?;
                let statement = Box::new(self.parse_statement()?);
                Statement::While {
                    pos,
                    condition,
                    statement,
                }
            }
            TokenKind::Repeat => {
                self.next();
                self.expect(&TokenKind::LeftParen)?;
                let expr = self.parse_expr()?;
                self.expect(&TokenKind::RightParen)?;
                let statement = Box::new(self.parse_statement()?);
                Statement::Repeat {
                    pos,
                    expr,
                    statement,
                }
            }
            TokenKind::Return => {
                self.next();
                let expr = self.parse_expr()?;
                self.expect(&TokenKind::Semicolon)?;
                Statement::Return { pos, expr }
            }
            TokenKind::Inline => {
                self.next();
                let token = self.next();
                match token.kind {
                    TokenKind::StringLiteral(code) => {
                        self.expect(&TokenKind::Semicolon)?;
                        Statement::Inline { pos, code }
                    }
                    _ => return Err(expected("string literal", token)),
                }
            }
            TokenKind::Type(_) => {
                let mut decl = self.parse_declaration()?;
                if self.optional(&TokenKind::Eq) {
                    decl.init = Some(self.parse_expr()?);
                }
                self.expect(&TokenKind::Semicolon)?;
                Statement::Declaration(decl)
            }
            TokenKind::Identifier(_) => {
                let name = self.expect_identifier()?;
                if self.optional(&TokenKind::LeftParen) {
                    // function call
                    let args = self.parse_list(Self::parse_expr, &TokenKind::RightParen)?;
                    self.expect(&TokenKind::Semicolon)?;
                    Statement::Call { name, args }
                } else {
                    // assignment
                    let token = self.next();
                    let pos = token.pos;
                    let kind = match token.kind {
                        TokenKind::Eq => AssignOpKind::Eq,
                        TokenKind::PlusEq => AssignOpKind::PlusEq,
                        TokenKind::MinusEq => AssignOpKind::MinusEq,
                        TokenKind::StarEq => AssignOpKind::StarEq,
                        TokenKind::SlashEq => AssignOpKind::SlashEq,
                        TokenKind::PercentEq => AssignOpKind::PercentEq,
                        _ => return Err(expected("function call or assignment", token)),
                    };
                    let op = AssignOp { pos, kind };
                    let expr = self.parse_expr()?;
                    self.expect(&TokenKind::Semicolon)?;
                    Statement::Assign { name, op, expr }
                }
            }
            _ => return Err(expected("statement", self.next())),
        })
    }

    fn parse_expr(&mut self) -> CompilerResult<Expr> {
        self.parse_expr_bp(0)
    }

    fn parse_expr_bp(&mut self, min_bp: u8) -> CompilerResult<Expr> {
        // pratt parsing
        // see https://matklad.github.io/2020/04/13/simple-but-powerful-pratt-parsing.html

        // prefix operators
        let mut lhs = if let Some((op, r_bp)) = unary_bp(&self.peek().kind) {
            let token = self.next();
            let expr = self.parse_expr_bp(r_bp)?;
            Expr::Unary {
                op: UnaryOp {
                    pos: token.pos,
                    kind: op,
                },
                right: Box::new(expr),
            }
        } else {
            self.parse_primary()?
        };

        // infix operators
        while let Some((op, (l_bp, r_bp))) = binary_bp(&self.peek().kind) {
            if l_bp < min_bp {
                break;
            }
            let op_token = self.next();
            let rhs = self.parse_expr_bp(r_bp)?;
            lhs = Expr::Binary {
                left: Box::new(lhs),
                op: BinaryOp {
                    pos: op_token.pos,
                    kind: op,
                },
                right: Box::new(rhs),
            };
        }

        Ok(lhs)
    }

    fn parse_primary(&mut self) -> CompilerResult<Expr> {
        let token = self.next();
        let pos = token.pos;
        Ok(match token.kind {
            TokenKind::Identifier(name) => {
                let name = Ident { pos, name };
                if self.optional(&TokenKind::LeftParen) {
                    // function call
                    let args = self.parse_list(Self::parse_expr, &TokenKind::RightParen)?;
                    Expr::Call { name, args }
                } else {
                    // variable
                    Expr::Var { name }
                }
            }
            TokenKind::LeftParen => {
                let expr = self.parse_expr()?;
                self.expect(&TokenKind::RightParen)?;
                expr
            }
            TokenKind::IntLiteral(value) | TokenKind::CharLiteral(value) => {
                Expr::Int { pos, value }
            }
            _ => return Err(expected("expression", token)),
        })
    }
}

pub fn parse_program(tokens: Vec<Token>, program_name: &str) -> CompilerResult<Program> {
    Parser::new(tokens.into_iter()).parse_program(program_name)
}
